### ISSUE TITLE
PROPOSAL: `CpiGuard` Extension

### ISSUE THREAD

**Issue by 2501babe** [solana-labs/solana-program-library#3694]
# CpiGuard

## tldr

cpi guard is an extension that prevents certain token actions inside cpi, to protect users from implicitly signing for actions they cant see, hidden in third-party programs

full summary of changes:
* changes to existing token-2022 processors:
  - `transfer`: if extension enabled and in cpi, authority *must* be delegate, or abort
  - `burn`: if extension enabled and in cpi, authority *must* be delegate, or abort
  - `close_account`: if extension enabled and in cpi, destination *must* be owner, or abort [IN FLUX: we may further limit to native mint accounts]
  - `approve`: if extension enabled and in cpi, abort
  - `set_authority`:
    * if `Account` and extension enabled and in cpi *and* authority to change is close authority *and* intent is to remove it, remove it
    * otherwise, if `Account` and extension enabled and in cpi, abort
    * if `Account` and authority to change is owner and extension enabled, abort *regardless of cpi* (this is to prevent moving a `CpiGuard` account to a program)
* new `CpiGuard` instruction and processor, modeled after `RequiredTransferMemos` interface. changing `CpiGuard` status in cpi is *not allowed*
* program tests for all invariants listed above
* token-client support for `CpiGuard`
* token-cli support for `CpiGuard` (todo after extension merged)
* token-js support for `CpiGuard` (todo after extension merged)
* documentation in `token-2022/extensions.mdx` for `CpiGuard` (todo after extension merged)

this is an extension that can be enabled or disabled by users at-will. it does not break backwards compat. rather, it simply automates a choice to systematically decline to sign certain types of transactions

## background

interacting with a dapp as a user means signing transactions that are constructed for you by its frontend code. there are three fundamental ways for a dapp to transfer funds from the user to the dapp (or, equivalently, burn them):
* insert a transfer instruction in the transaction, which the user signs for
* insert an approve instruction in the transaction, which the user signs for, and perform a cpi transfer under program authority
* require the token program and a user signer for an opaque program instruction, and perform a cpi transfer under user authority

the first two are safe, in that the user can see exactly what is being done, with zero ambiguity. the third is quite dangerous. passing in a wallet signature allows the program to perform any action the user could, without any visibility into what it may be doing. there have been some attempts at workarounds, for instance phantom may simulate the transaction and attempt to warn about balance changes. but, fundamentally, this is intractable [jocular citation of turing 1936]

this also doesnt cover the approve and set authority instructions, which can grant authority to transfer all user funds without immediately effecting a balance change. wallets could also warn if the various authorities change after simulation, but again, this isnt perfect

there are two ways to make this much safer that _dont_ have to grapple with undecidability:
* wallets should warn whenever a wallet signature is made available to an opaque (non-system, non-token) instruction, and users should be educated to treat the request for a signature on such an instruction as highly suspect
* the token program should have a means to prevent cpi calls on user authority from taking place, forcing opaque programs to directly ask for such authority in a user-visible manner

this proposal concerns the second

## proposal

add a new `Account` extension to Token2022: `CpiGuard`

this would allow a user to _opt in_ to backwards-compat-breaking behavior, rejecting certain kinds of cpi actions involving their token account:
* `transfer` via cpi must go through a delegate
* `burn` via cpi must go through a delegate
* `close_account` via cpi must go through a close authority
* `approve` via cpi is banned
* `set_authority` via cpi is banned
* disabling cpi guard via cpi is banned

this change is _non-breaking_ for users who do not enable it, and users are free to enable or disable it as they please. the purpose of this is to protect users from accidentally granting a program permission to perform actions with their signature, if they personally choose not to allow such behavior

it also is part of an attempt to nudge program authors toward designing with "nothing up their sleeve," creating a ux pitfall that encourages them to use user-authed token instructions rather than user-authed cpi calls. especially with address maps live and double-size transactions in the pipeline, program authors should have sufficient space in their transactions to do things the right way

this change does _not_ impact the ability of programs to auth their _own_ accounts, or accounts of other programs, via cpi. they should simply not enable the extension

## thoughts concerns etc

i waffled on whether to ban `close_account` entirely or not. i decided against it because, while close authority does behave in a surprising manner, the user is free to grant it to a program if they really want to

we might want to ban changing cpi guard status in cpi entirely, because if a program enabled it for their own account, they would not be able to disable it. there is also a very stupid edge case where a user could transfer ownership of a cpi-locked account to a program, which then wouldnt be able to close it. honestly i realized _while writing this proposal_ that we needed to ban removing cpi guard in cpi or else a program could trivially bypass it

the tricky bit here is we cannot detect whether an account is a pda or an eoa without an ed25519 syscall, and requiring this overcomplicates a relatively simple extension. perhaps we unconditionally ban changing cpi guard status inside cpi, and we unconditionally ban transferring ownership of an account unless cpi guard is disabled. this feels right but i havent fully thought through all the edge cases

i considered making this extension trimodal, with a third "ban token actions via cpi entirely" state. but this seems superfluous. the point is not to prevent action from being taken, but to force developers to expose those actions to the user

its worth noting that this may seem like it overlaps the functionality of `ImmutableOwner`, but it doesnt. we prevent a program from changing an account owner via cpi, but the program is still free to surface a `set_authority` instruction for owner, and a user is free to sign for it

the owner/authority distinction is not perfect, if the user were for whatever reason to set themself as their own delegate or close authority (assuming this isnt blocked the processor, i havent checked). we could check that the supposed authority is different from owner. but we could also just say this is dumb

## alternate names

my original name was `SafeCpi`, which i immediately discarded in favor of `SaferCpi` so as to disclaim all liability for the consequences of my actions. `CpiLock` was another candidate. `TokenGuard` or `NoSignaturePassthrough` or some other things were also considered. i think `CpiGuard` sounds best, but im open to suggestions

## fin

please let me know questions/comments/words of encouragment/stern rebukes and chastisements, and feel free to tag anyone who you think might also have thoughts or should also have visibility on this. ideally i get this in before we freeze Token2022, but, i guess we will see

cc @joncinque @CriesofCarrots @mvines

**Comment by joncinque** [comment]
This all looks great. The workings of it are very clear, and I agree that it's best to lock down as much as possible for full safety.  This will absolutely break things, but it's easy afterwards to go to a dapp and say "hey, why don't you support safer accounts?" and force them to do the right thing.

For this point in particular:

> * close_account via cpi must go through a close authority

This one can cause a bigger footgun in the future.  For delegates, it's easy to create an ephemeral keypair for one transaction, lose the key, and then just revoke it later.  With a separate close authority, you're forced to keep that key around afterwards, so what key will a wallet use?  Would you have to sandwich your call to the program with `set_authority` on both sides to set it to an ephemeral keypair and set it back? If you forget the second one, you've lost close authority on that account for good.

On the flip side, if there's no guard on `close`, sketchy dapps can use your signature on a wrapped SOL account to take all the SOL in it.  This is pretty specific, but here's my idea:

* close_account via CPI must set the owner as the recipient of the lamports
 
This way, a sketchy program can't do anything nasty with your signature, which seems to be the main intent of this extension.

**Comment by 2501babe** [comment]
>With a separate close authority, you're forced to keep that key around afterwards
theres no need for an eoa in this case. it would look like: `set_authority` instruction grants close authority to a pda. program can close the account if it chooses

i was gonna say "or repudiate its authority" but now i realize that if we ban `set_authority` through cpi, the user would need to turn off `CpiGuard`. but at least it wouldnt require their signature? mm, i guess we could allow `set_authority` via cpi if its _removing_ permissions rather than _adding_ them, but this adds code complexity (attack surface)

hmm i guess letting a program call `close_account` isnt _too_ bad since its already required to be empty of tokens...

>This will absolutely break things
i feel compelled to repeat (for the benefit of onlookers) that this extension is opt-in and does not break backwards compat. if disabled, nothing different happens. and if enabled, it is equivalent to a user carefully reading every transaction and choosing not to sign it if it does particular things they dont personally approve of. the "breakage" to dapp authors is "i create a suboptimal ux experience for discerning consumers," not "certain behaviors that were possible before become impossible"

**Comment by 2501babe** [comment]
unrelated, i had the thought that maybe `ImmutableOwner` should prevent setting a close authority

**Comment by joncinque** [comment]
> it would look like: set_authority instruction grants close authority to a pda. program can close the account if it chooses

Maybe I'm missing something.  When a user grants close authority to another address, they can't get it back unless that address gives it back.  Is the assumption that a user would only do that for a good reason? ie. they trust the program to do the right thing.  But that goes against the central tenet of this extension, which is that we *don't* trust the program.

> hmm i guess letting a program call close_account isnt too bad since its already required to be empty of tokens

The case I'm worried about is wrapped SOL, where all of the SOL in the account gets sent to the destination. If I sign something and pass my wrapped SOL account, a sketchy program can still steal all of the SOL in that account.

> i feel compelled to repeat (for the benefit of onlookers) that this extension is opt-in and does not break backwards compat

Sorry, to be totally clear, what I meant is: it will break things in beautiful and good ways when enabled.

**Comment by 2501babe** [comment]
>But that goes against the central tenet of this extension, which is that we don't trust the program.

i would say the central tenet is more like, the program must ask for the permissions in a way that allows the user to knowingly and willingly grant them. if set authority took place _via cpi_, it would be a violation of that intent. we ban that case. but if set authority takes place _as a visible token instruction_, the user is free to do whatever stupid thing they please, just as a user is free to delegate all their tokens to a program

but the fact that the user cannot revoke close authority if they change their mind does make it different, and i agree we shouldnt be incentivizing programs to ever use this feature. i think youre right about allowing close authority but returning lamports to owner

**Comment by 2501babe** [comment]
this is my current plan, in more programmer-y (ie, making it clear what branches i need to add) format:

* changes to existing processors:
  - `transfer`: if extension enabled and cpi, authority *must* be delegate, or abort
  - `burn`: if extension enabled and cpi, authority *must* be delegate, or abort
  - `close_account`: if extension enabled and cpi, destination *must* be owner, or abort
  - `approve`: if extension enabled and cpi, abort
  - `set_authority`:
    * if `Account` and extension enabled and cpi *and* authority to change is close authority *and* intent is to remove it, remove it
    * otherwise, if `Account` and extension enabled and cpi, abort
    * if `Account` and authority to change is owner and extension enabled, abort *regardless of cpi* (this is to prevent moving a `CpiGuard` account to a program)
* new `CpiGuard` instruction and processor, modeled after `RequiredTransferMemos` interface. changing `CpiGuard` status in cpi is *not allowed*
* program tests for all invariants listed above
* token-client support for `CpiGuard` (todo after program finalized)
* token-cli support for `CpiGuard` (todo after program finalized)

**Comment by 2501babe** [comment]
ok heres the baldest yak ive ever encountered. in doc comments, do we want to refer to the act of calling into a program like:
* cpi (noun), eg "This happens in cpi."
* Cpi (proper noun), eg "This happens in Cpi."
* cross-program (adjective), eg "This happens cross-program." or "This happens in a cross-program context."
* Cross-Program Invocation (proper noun), eg "This happens via Cross-Program Invocation."

i like 1, 2 feels more "correct" but also somewhat strange because unlike most other proper nouns here (Account, Transfer, Required Memo Transfers) it refers to an operating environment, rather than a concrete struct or enum variant

**Comment by joncinque** [comment]
Looks like it's officially called "CPI" or "cross-program invocation" https://docs.solana.com/terminology#cross-program-invocation-cpi

**Comment by 2501babe** [comment]
ah also i grepped the monorepo and CPI is ubiquitous but Cpi nonexistent. im gonna follow the convention we use with bpf: BPF in documentation, Bpf when part of type nams

**Comment by nahnicholas** [comment]
Ty
